storage-operator/cmd/manager/main.go

```diff
@@ -116,39 +161,47 @@ func main() {
 		{Port: metricsPort, Name: metrics.OperatorPortName, Protocol: v1.ProtocolTCP, TargetPort: intstr.IntOrString{Type: intstr.Int, IntVal: metricsPort}},
 		{Port: operatorMetricsPort, Name: metrics.CRPortName, Protocol: v1.ProtocolTCP, TargetPort: intstr.IntOrString{Type: intstr.Int, IntVal: operatorMetricsPort}},
 	}
-	// Create Service object to expose the metrics port(s).
-	ctx := context.TODO()
-	_, err = metrics.CreateMetricsService(ctx, cfg, servicePorts)
 
+	// Create Service object to expose the metrics port(s).
+	service, err := metrics.CreateMetricsService(ctx, cfg, servicePorts)
 	if err != nil {
-		log.Info(err.Error())
+		log.Info("Could not create metrics Service", "error", err.Error())
 	}
 
-	log.Info("Starting the Cmd.")
+	// CreateServiceMonitors will automatically create the prometheus-operator ServiceMonitor resources
+	// necessary to configure Prometheus to scrape metrics from this operator.
+	services := []*v1.Service{service}
 
-	// Start the Cmd
-	if err := mgr.Start(signals.SetupSignalHandler()); err != nil {
-		log.Error(err, "Manager exited non-zero")
-		os.Exit(1)
+	// The ServiceMonitor is created in the same namespace where the operator is deployed
```

Comment:
Unfortunately, those ServiceMonitors won't be picked up by our prometheus-operator (since it uses the `release=prometheus-operator` label selector, see #2199)... Looked quickly at their code, not sure we can inject custom labels, the `metrics.CreateServiceMonitors` relies on a custom method on `v1.Service` instance called [`GenerateServiceMonitor`](https://github.com/operator-framework/operator-sdk/blob/master/pkg/metrics/service-monitor.go#L73-L104), which basically copies the Service labels for the ServiceMonitor... And the Service created by `metrics.CreateMetricsService` only has one label: `name=<operator-name>`.

So I guess we'd need to patch the ServiceMonitor(s) after creation to add those labels if we wanted a workaround, and probably should open an issue on Operator SDK's tracker.

Reply:
Well, `CreateServiceMonitors` takes optional functions to modify the `ServiceMonitor` object before it being created (https://github.com/operator-framework/operator-sdk/blob/0fa4c824583d60fab2f5c0de3275d6a8bde56982/pkg/metrics/service-monitor.go#L37). So, adding this `release=prometheus-operator` label to them should fix this, right?

Reply:
Now, in general, I'm not that fond of the *operator* creating its own `ServiceMonitor`, `Service` and whatnot objects *at runtime*.

Reply:
In any case, this should not 'block' this PR:

- The same issue exists with the 'old' code, so metrics aren't captured today.
- We wouldn't be doing anything *useful* with these metrics anyway for now.
- Solution (whilst feasible to implement, cfr. comment above) would better wait for #2199 instead of now patching in the `release=prometheus-operator` label
- I'd rather 'deploy a `ServiceMonitor` with the operator deployment' and get rid of this 'runtime generation'.

Reply:
Didn't catch the `updaters` argument! So we're fine :)

> I'd rather 'deploy a ServiceMonitor with the operator deployment' and get rid of this 'runtime generation'

Well, I kind of agree in a sense, but also like the idea of considering the Operator as a self-bootstrapping installer... For this specific Operator, it doesn't change much because we deploy it with Salt, so still nothing to do for the user, but for an Operator in the general sense, I understand why it's the default.